lib/Optimizer/Transforms/StatePreparation.cpp

```diff
@@ -0,0 +1,209 @@
+///     %1 = quake.alloca !quake.veq<2>
+///     %2 = quake.init_state %1, %0 : (!quake.veq<2>,
+///       !cc.ptr<!cc.array<complex<f32> x 4>>) -> !quake.veq<2> return
+///  }
+///  cc.global constant @foo.rodata_0 (dense<[(0.707106769,0.000000e+00),
+///      (0.707106769,0.000000e+00), (0.000000e+00,0.000000e+00),
+///      (0.000000e+00,0.000000e+00)]> : tensor<4xcomplex<f32>>) :
+///    !cc.array<complex<f32> x 4>
+/// }
+///
+/// After StatePreparation (state-prep):
+///
+/// module {
+///   func.func @foo() attributes {
+///     %0 = quake.alloca !quake.veq<2>
+///     %c1_i64 = arith.constant 1 : i64
+///     %1 = quake.extract_ref %0[%c1_i64] : (!quake.veq<2>, i64) -> !quake.ref
+///     %cst = arith.constant 0.000000e+00 : f64
+///     quake.ry (%cst) %1 : (f64, !quake.ref) -> ()
+///     %c0_i64 = arith.constant 0 : i64
+///     %2 = quake.extract_ref %0[%c0_i64] : (!quake.veq<2>, i64) -> !quake.ref
+///     %cst_0 = arith.constant 0.78539816339744839 : f64
+///     quake.ry (%cst_0) %2 : (f64, !quake.ref) -> ()
+///     quake.x [%1] %2 : (!quake.ref, !quake.ref) -> ()
+///     %cst_1 = arith.constant 0.78539816339744839 : f64
+///     quake.ry (%cst_1) %2 : (f64, !quake.ref) -> ()
+///     quake.x [%1] %2 : (!quake.ref, !quake.ref) -> ()
+///     return
+///   }
+/// }
+
+namespace {
+
+std::vector<std::complex<double>>
+readConstantArray(mlir::OpBuilder &builder, cudaq::cc::GlobalOp &global) {
+  std::vector<std::complex<double>> result{};
+
+  auto attr = global.getValue();
+  auto type = global.getType().getElementType();
+
+  auto arrayTy = dyn_cast<cudaq::cc::ArrayType>(type);
+  assert(arrayTy);
+  assert(attr.has_value());
+
+  auto elementsAttr = dyn_cast<mlir::ElementsAttr>(attr.value());
+  assert(elementsAttr);
+  auto eleTy = elementsAttr.getElementType();
+  auto values = elementsAttr.getValues<mlir::Attribute>();
+
+  for (auto it = values.begin(); it != values.end(); ++it) {
+    auto valAttr = *it;
+
+    auto v = [&]() -> std::complex<double> {
+      if (isa<FloatType>(eleTy))
+        return {cast<FloatAttr>(valAttr).getValue().convertToDouble(),
+                static_cast<double>(0.0)};
+      if (isa<IntegerType>(eleTy))
+        return {static_cast<double>(cast<IntegerAttr>(valAttr).getInt()),
+                static_cast<double>(0.0)};
+      assert(isa<ComplexType>(eleTy));
+      auto arrayAttr = cast<mlir::ArrayAttr>(valAttr);
+      auto real = cast<FloatAttr>(arrayAttr[0]).getValue().convertToDouble();
+      auto imag = cast<FloatAttr>(arrayAttr[1]).getValue().convertToDouble();
+      return {real, imag};
+    }();
+
+    result.push_back(v);
+  }
+  return result;
+}
+
+LogicalResult transform(ModuleOp module, func::FuncOp funcOp) {
+  auto builder = OpBuilder::atBlockBegin(&funcOp.getBody().front());
+  auto toErase = std::vector<mlir::Operation *>();
+  auto hasInitState = false;
+  auto replacedInitState = false;
+
+  funcOp->walk([&](Operation *op) {
```

Comment:
Why are we using `walk` and building a state machine here?

Reply:
Some state machine is needed - the pass succeeds if all init state instructions are removed, so I need to track it somehow. I modified the code to be less state-machine-like, let me know if it looks better.

Reply:
Letting this go for now, but let's open an issue.

We should be able to write a partial conversion that will necessarily eliminate all instances of whatever operations we tell it we want eliminated. It will remove the boilerplate, apply simplifications, etc. all under the hood for us.